### PR TITLE
OKR add feature : upload PDF

### PR DESCRIPTION
**Pull request by antonio-do**

### CODE REVIEW

**Comment by antonio-do** [comment]
Monday review @ryo-suwito 
- [ ] You need to have text to tell users that they can upload files after they create OKR
- [ ] You need a confirmation for deleting file

**Comment by antonio-do** [comment]
Ryo ask: are you using Nginx for files handing?
- No: we are using http for uploading file, because we want to organize file in your own way
- Yes: we could use Nginx for downloading

File: okr_app/tests.py [hunk 1]
```diff
@@ -1,4 +1,100 @@
...
+        self.user.save()
+        self.userGuest = User.objects.create(username='guest')
+        self.userGuest.set_password('12345')
+        self.userGuest.save()
+        self.okr = OKR(issuer=self.user, quarter="1", year="2021", content="test content")
+        self.okr.save()
+        self.okr_file = OKRFile(issuer=self.user, file=fileSample, okr=self.okr)
+        self.okr_file.save()
+
+    #Can OKR issuer get list of all his/her okr files?
```
Comment (antonio-do) [inline okr_app/tests.py:24]: using affirmative to describe test case, do not use interrogative, same for the others

Comment (antonio-do) [reply okr_app/tests.py:24]: put description in side the function with """ """
```
e.g:
def testcase(self):
    """testcase
    """
```

File: okr_app/views.py [hunk 2]
```diff
@@ -42,3 +45,37 @@ def update(self, request, *args, **kwargs):
...
+        #If request user is either superadmin or okr_admin, give them all
+        if is_admin:
+            return self.queryset
+        #Else return user's owned file only
+        return self.queryset.filter(okr__issuer=self.request.user)
+
+    def create(self, request, *args, **kwargs):
+        # Just in case we need a list of files uploaded by user
+        request.data['issuer'] = request.user.username
+        return super().create(request, *args, **kwargs)
```
Comment (antonio-do) [inline okr_app/views.py:80]:
```
user = self.request.user
okr = OKR.objects.get(pk=self.kwargs['okr'])
if IsApplicationAdminUser(user) or IsOKROwner(user, okr):
      return self.queryset.filter(okr_id=okr.id)

return self.queryset.none()
```

Comment (antonio-do) [reply okr_app/views.py:80]: listen to my instruction. Use this logic for queryset, then tell me what the difference between my code and your code, and find out by yourself where you were wrong.

File: okr_app/permissions.py [hunk 3]
```diff
@@ -14,3 +14,8 @@ def has_permission(self, request, view):
             return True
 
         return False
+
+class IsOKROwner(BasePermission):
+    # for object level permissions
+    def has_object_permission(self, user, okr):
```
Comment (antonio-do) [inline okr_app/permissions.py:20]: `def has_object_permission(self, request, view, obj):`

File: okr_app/tests.py [hunk 4]
```diff
@@ -1,4 +1,134 @@
...
+
+    #Can ?
+    def test_delete_with_okr_owner(self):
+        """
+        OKR issuer can delete OKR's uploaded file
+        """
+        self.client.login(username='okrowner', password='12345')
+        self.url = '/api/okrfiles/{id}/'.format(id=self.okr_file.id) 
+        response = self.client.delete(self.url, {}, format='json')
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
```
Comment (antonio-do) [inline okr_app/tests.py:134]: use coverage to make sure that your tests cover all the code
https://coverage.readthedocs.io/en/6.2/

File: okr_app/views.py [hunk 5]
```diff
@@ -42,3 +45,27 @@ def update(self, request, *args, **kwargs):
...
+        okr = None
+        if self.request.GET.get("okr", None):
+            okr = get_object_or_404(OKR, pk=self.request.GET.get("okr", None))
+        elif self.kwargs.get("pk", None):
+            okr = get_object_or_404(OKR, files=self.kwargs['pk'])
+
+        if okr and (IsApplicationAdminUser().has_permission(self.request, self) or IsOKROwner().has_object_permission(user, okr)):
+            return self.queryset.filter(okr_id=okr.id)
+
+        return self.queryset.none()
```
Comment (antonio-do) [inline okr_app/views.py:71]: test the case that a regular user uploads files for OKR that doesn't belong to him/her

File: okr_app/tests.py (hunk 4 above)
Comment (ryo-suwito) [reply okr_app/tests.py:134]: coverage

File: frontend/src/okr/actions/url.js [hunk 6]
```diff
@@ -1,8 +1,15 @@
 const OKRURL = (action = '') => (id = '') =>
   id ? (action ? `/api/okrs/${id}/${action}/` : `/api/okrs/${id}/`) : '/api/okrs/';
-
+const OKRFILEURL = (action = '') => (id = '') =>
+  id ? `/api/okrfiles/?okr=${id}` : '/api/okrfiles/';
+const OKRFILEDELETEURL = (action = '') => (id = '') =>
+  `/api/okrfiles/${id}/`;
```
Comment (antonio-do) [inline frontend/src/okr/actions/url.js:6]: `action` is never used in the functions. You should rather remove it.

File: frontend/src/okr/actions/constants.js [hunk 7]
```diff
@@ -5,4 +5,9 @@ export const API_MESSAGES = {
   ON_CREATE_OKR: "You've created OKR successfully!",
   ON_CREATE_OKR_ERROR: "Error",
   ON_DELETE_OKR: "You've deleted OKR successfully!",
+
+  ON_CREATE_OKR_FILE: "You've created OKR File successfully!",
+  ON_CREATE_OKR_FILE_ERROR: "Error",
+  ON_DELETE_OKR_FILE: "You've deleted OKR File successfully!",
+  ON_DELETE_OKR_FILE_ERROR: "Error Deleting OKR File",
```
Comment (antonio-do) [inline frontend/src/okr/actions/constants.js:12]: shorten the noun "OKR File" -> "file"

**Pull request merged**